src/bootstrap/builder.rs

```diff
@@ -0,0 +1,512 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::cell::RefCell;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+use std::fs;
+use std::ops::Deref;
+use std::any::{TypeId, Any};
+
+use compile;
+use install;
+use dist;
+use util::{exe, libdir, add_lib_path};
+use {Build, Mode};
+use cache::{INTERNER, Interned, Cache};
+use check;
+use flags::Subcommand;
+use doc;
+use tool;
+
+pub use Compiler;
+
+pub struct Builder<'a> {
+    pub build: &'a Build,
+    pub top_stage: u32,
+    pub kind: Kind,
+    cache: Cache,
+    stack: RefCell<Vec<(TypeId, Box<Any>)>>,
```

Comment:
Yeah I think it can go, you'll just need to fiddle with a few conditions in `ensure`. Will probably end up simpler as well.